src/simperium/bucket.js

```diff
@@ -1,60 +1,435 @@
-import { EventEmitter } from 'events'
-import { inherits } from 'util'
-import uuid from 'node-uuid';
-
-export default function Bucket( name, storeProvider ) {
-	EventEmitter.call( this );
-	this.name = name;
-	this.store = storeProvider( this );
-	this.isIndexing = false;
-}
+// @flow
+import events from 'events'
+import uuid from 'uuid/v4';
 
-inherits( Bucket, EventEmitter );
+const { EventEmitter } = events;
 
-Bucket.prototype.reload = function() {
-	this.emit( 'reload' );
-};
+type TaskCallback<T> = ( ?Error, ?T ) => void;
 
-Bucket.prototype.add = function( object, callback ) {
-	var id = uuid.v4();
-	return this.update( id, object, callback );
+/**
+ * Runs a promise with a callback (if one is provided) to support the old callback API.
+ * NOTE: if the callback API is removed this is a place to warn users
+ *
+ * @param {Function} [callback] - if provided, will be called with the expected values
+ * @param {Promise} promise - promise to run, executes callback if provieded
+ * @returns {Promise} promise is passed through
+ */
+const deprecateCallback = <T>( callback: ?TaskCallback<T>, promise: Promise<T> ): Promise<T> => {
+	if ( callback ) {
+		const perform = callback;
+		// Potentially could warn here if we decide to remove callback API
+		return promise.then(
+			( result ) => {
+				perform( null, result );
+				return result;
+			},
+			( error ) => {
+				perform( error, null );
+				return error;
+			}
+		);
+	}
+	return promise;
 };
 
-Bucket.prototype.get = function( id, callback ) {
-	return this.store.get( id, callback );
+export type BucketObject = {
+	id: string,
+	data: {},
+	isIndexing?: boolean
+}
+
+export interface BucketStore {
+	/**
+	 * Retrieve a bucket object from the store
+	 * @function
+	 * @name BucketStore#get
+	 * @param {String} id - the bucket object id to fetch
+	 * @param {bucketStoreGetCallback} - callback once the object is fetched
+	 */
+	get( id: string, callback: ( ?Error, ?BucketObject ) => void ): void;
```

Comment:
why not standardize on a Promise interface? that would let people more easily use `async/await` too

Reply:
Thats the goal, this is just keeping backwards compatibility right now.